File: obeltests/histogram.m
//
//  histogram.m
//  obeltests
//
//  Created by Jack Mousseau on 9/23/21.
//

#import <obel/histogram.h>
#import <XCTest/XCTest.h>

@interface ObelHistogramTests : XCTestCase

@end

@implementation ObelHistogramTests

- (void)testHistogramBin {
    {
        XCTAssertEqual(obel_histogram_bin(0.f, 4.f, 28.f), 7);
        XCTAssertEqual(obel_histogram_bin(1.f, 4.f, 28.f), 6);
    } {
        int32_t bins[10] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
        obel_histogram_t histogram = { .offset = 0.f, .stride = 4.f, .count = 10, .samples = 0, .bins = bins };
        XCTAssertEqual(obel_histogram_bin(&histogram, 28.f), 7);
        XCTAssertEqual(histogram.bins[7], 1);
        XCTAssertEqual(histogram.samples, 1);
        histogram.offset = 1.f;
        XCTAssertEqual(obel_histogram_bin(&histogram, 26.f), 6);
        XCTAssertEqual(obel_histogram_bin(&histogram, 27.f), 6);
        XCTAssertEqual(obel_histogram_bin(&histogram, 28.f), 6);
        XCTAssertEqual(histogram.bins[6], 3);
    }
}

- (void)testHistogramDistance {
    int32_t bins_a[4] = { 5, 2, 6, 4 };
    int32_t bins_b[4] = { 1, 7, 3, 8 };
    obel_histogram_t histogram_a = { .offset = 0.f, .stride = 1.f, .count = 4, .samples = 16, .bins = bins_a };
    obel_histogram_t histogram_b = { .offset = 0.f, .stride = 1.f, .count = 4, .samples = 19, .bins = bins_b };
    XCTAssertEqual(obel_histogram_distance(&histogram_a, &histogram_b), 16);
}

@end

File: obel/obel/histogram.c
//
//  histogram.c
//  obel
//
//  Created by Jack Mousseau on 9/22/21.
//

#include <obel/histogram.h>

OBEL_HISTOGRAM_FUNC int32_t obel_histogram_bin(obel_histogram_t *histogram, float element) {
    return obel_histogram_bin(histogram->offset, histogram->stride, element);
}

OBEL_HISTOGRAM_FUNC int32_t obel_histogram_bin(float offset, float stride, float element) {
    return (int32_t)floorf((element - offset) / stride);
}

OBEL_HISTOGRAM_FUNC int32_t obel_histogram_distance(obel_histogram_t *a, obel_histogram_t *b) {
    assert(a->offset == b->offset);
    assert(a->stride == b->stride);
    assert(a->count == b->count);

    int32_t distance = 0;
    for (int32_t i = 0; i < a->count; i++) {
        distance += abs(a->bins[i] - b->bins[i]);
    }

    return distance;
}
